numpy/_core/src/multiarray/item_selection.c

```diff
@@ -1067,7 +1069,18 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
             flags |= NPY_ARRAY_ENSURECOPY;
         }
         Py_INCREF(dtype);
-        obj = (PyArrayObject *)PyArray_FromArray(out, dtype, flags);
+        if (!PyArray_EquivTypes(dtype, PyArray_DESCR(out))) {
+            obj = (PyArrayObject *)PyArray_NewFromDescr(Py_TYPE(out),
+                                                        dtype,
+                                                        multi->nd,
+                                                        multi->dimensions,
+                                                        NULL, NULL, 0,
+                                                        (PyObject *)out);
+            is_newarr = 1;
+        }
+        else {
+            obj = (PyArrayObject *)PyArray_FromArray(out, dtype, flags);
+        }
     }
 
     if (obj == NULL) {
```

Comment:
Can't comment lower, but we should adjust the `PyArray_GetDTypeTransferFunction` to use `PyArray_DESCR(obj)` for the second dtype.

There is a subtle thing that i.e. `StringDType` can be equivalent but _not_ identical (which is totally fine!).
(I.e. this code _should_ currently lead to crashes if used on the new `StringDType`, I believe.  New tests for that would be awesome, but I am happy to overlook that.)

Reply:
I adjusted the function, and added a test for `StringDType`, which does not crash now!

Reply:
Honestly, I thought this may be a new issue.  But it is clearly an existing bug and maybe the best thing is to just leave it out here.

I suppose `CastToCommonType` is buggy, in that it should ensure identical (view-castable), which we have a helper for now.
(I.e. replacing an `EquivTypes` with the new helper, but I forgot that helpers name :))